File: iOS/WATisRain/WATisRain/MapView.m

#import <Foundation/Foundation.h>
#import "MapView.h"
#import "MapFactory.h"
#import "Building.h"
#import "Map.h"
#import "Util.h"

@implementation MapView

- (MapView*)initWithImage:(UIImage*)image{
    _image = image;
    self = [super initWithFrame:CGRectMake(0, 0, image.size.width, image.size.height)];
    
    _defaultLocationImage = [UIImage imageNamed:@"default_location.png"];
    _activeLocationImage = [UIImage imageNamed:@"active_location.png"];
    _stairsUpImage = [UIImage imageNamed:@"stairs_up.png"];
    _stairsDownImage = [UIImage imageNamed:@"stairs_down.png"];
    
    NSString *path = [[NSBundle mainBundle] pathForResource:@"locations" ofType:@"txt" inDirectory:@""];
    _map = [MapFactory readMapFromPath:path];
    _routeFinder = [[RouteFinder alloc] initWithMap:_map];
    
    return self;
}

- (void)handleUserTapOnX:(float)x OnY:(float)y{
    _clearBtn.enabled = true;
    _clearBtn.title = @"Clear";
    
    float map_x = (x / MAP_ADJUST_SCALING) + MAP_ADJUST_X;
    float map_y = (y / MAP_ADJUST_SCALING) + MAP_ADJUST_Y;
    
    Building *closestBuilding = [self determineBuildingFromPositionWithX:map_x WithY:map_y Threshold:70];
    if(closestBuilding == nil ||
       [[closestBuilding name] isEqual:_selectedBuilding1] ||
       [[closestBuilding name] isEqual:_selectedBuilding2]){
        [self clearRoute];
    }
    else if(_selectedBuilding1 == nil){
        _selectedBuilding1 = [closestBuilding name];
        [_directionsView selectDestination:_selectedBuilding1];
    }else{
        _selectedBuilding2 = [closestBuilding name];
        [self updateRoute];
    }

}

- (void)clearRoute{
    _selectedBuilding1 = nil;
    _selectedBuilding2 = nil;
    _route = nil;
    [_directionsView unselectDestination];
    _clearBtn.enabled = false;
    _clearBtn.title = @"";
    [self setNeedsDisplay];
}

- (void)updateRoute{
    _route = [[_routeFinder findRouteFrom:[_map getBuildingByID:_selectedBuilding1] To:[_map getBuildingByID:_selectedBuilding2]] getContractedRoute];
    [_directionsView generateDirectionsFromRoute:_route];
}

- (void)recalculateRoute{
    if(_selectedBuilding1 == nil || _selectedBuilding2 == nil) return;
    [self updateRoute];
    [self setNeedsDisplay];
}

void drawImageOnMap(CGContextRef context, UIImage *img, float x_, float y_, float w_){
    float x = (x_ - MAP_ADJUST_X) * MAP_ADJUST_SCALING;
    float y = (y_ - MAP_ADJUST_Y) * MAP_ADJUST_SCALING;
    float w = w_ * MAP_ADJUST_SCALING;
    
    [img drawInRect:CGRectMake(x-w/2.0, y-w/2.0, w, w)];
}

void drawPathOnMap(CGContextRef context, Path *path){
    NSArray *wps = [path waypoints];
    
    CGContextSetStrokeColorWithColor(context, [UIColor colorWithRed:0.0f green:0x70/255.0f blue:0xcf/255.0f alpha:1.0f].CGColor);
    CGContextSetRGBFillColor(context, 0.0f, 0x70/255.0, 0xcf/255.0, 1.0f);
    CGContextSetLineWidth(context, 8);
    for(int i=0; i<[wps count]-1; i++){
        Waypoint *wp1 = [wps objectAtIndex:i];
        Waypoint *wp2 = [wps objectAtIndex:i+1];
        float xx1 = ([wp1 x] - MAP_ADJUST_X) * MAP_ADJUST_SCALING;
        float yy1 = ([wp1 y] - MAP_ADJUST_Y) * MAP_ADJUST_SCALING;
        float xx2 = ([wp2 x] - MAP_ADJUST_X) * MAP_ADJUST_SCALING;
        float yy2 = ([wp2 y] - MAP_ADJUST_Y) * MAP_ADJUST_SCALING;
        CGContextMoveToPoint(context, xx1, yy1);
        CGContextAddLineToPoint(context, xx2, yy2);
        CGContextStrokePath(context);
        
        // draw caps
        CGContextFillEllipseInRect(context, CGRectMake(xx1-4, yy1-4, 8, 8));
        CGContextFillEllipseInRect(context, CGRectMake(xx2-4, yy2-4, 8, 8));
    }

}

- (void)drawRect:(CGRect)rect{
    CGContextRef context = UIGraphicsGetCurrentContext();
    [self.image drawAtPoint:CGPointMake(0, 0)];

    // Draw all locations
    for(Building *building in [_map buildings]){
        if(![building selectable]) continue;
        drawImageOnMap(context, _defaultLocationImage, [[[building getMainFloor] position] x], [[[building getMainFloor] position] y], 120);
    }
    
    // Draw route
    if(_route != nil){
        NSArray *all_steps = [_route routeSteps];
        for(RouteStep *step in all_steps){
            drawPathOnMap(context, [step path]);
        }
    }

    
    // Draw active locations
    for(Building *building in [_map buildings]){
        if([[building name] isEqual:_selectedBuilding1] ||
           [[building name] isEqual:_selectedBuilding2]){
            drawImageOnMap(context, _activeLocationImage, [[[building getMainFloor] position] x], [[[building getMainFloor] position] y], 120);
        }
    }
    
    // Draw stairs
    if(_route != nil){
        NSMutableArray *all_steps = [_route routeSteps];
        
        NSMutableArray *throughBuildings = [[NSMutableArray alloc] init];
        RouteStep *allsteps0 = [all_steps objectAtIndex:0];
        [throughBuildings addObject:[[allsteps0 start] getBuildingName]];
        for(RouteStep *step in all_steps) {
            NSString *next_build = [[step end] getBuildingName];
            if(![throughBuildings containsObject:next_build])
                [throughBuildings addObject:next_build];
        }
        
        NSMutableArray *throughWaypointsM = [[NSMutableArray alloc] init];
        for(RouteStep *step in all_steps){
            [throughWaypointsM addObjectsFromArray:[step getWaypoints]];
        }
        // remove duplicates
        NSMutableArray *throughWaypoints = [[NSMutableArray alloc] init];
        for(Waypoint *w in throughWaypointsM){
            if(![throughWaypoints containsObject:w])
                [throughWaypoints addObject:w];
        }
        
        for(NSString *buildingName in throughBuildings){
            int ix = [throughWaypoints indexOfObject:[[_map getBuildingByID:buildingName] position]];
            
            Waypoint *wp_cur = [throughWaypoints objectAtIndex:ix];
            Waypoint *wp_before;
            if(ix==0){
                wp_before = [throughWaypoints objectAtIndex:1];
            }else{
                wp_before = [throughWaypoints objectAtIndex:ix-1];
            }
            Waypoint *wp_after;
            if(ix==[throughWaypoints count]-1){
                wp_after = [throughWaypoints objectAtIndex:[throughWaypoints count]-2];
            }else{
                wp_after = [throughWaypoints objectAtIndex:ix+1];
            }
            
            NSArray *vec_c = findOppositeVector([wp_before x]-[wp_cur x], [wp_before y]-[wp_cur y], [wp_after x]-[wp_cur x], [wp_after y]-[wp_cur y]);
            
            BOOL stairs_down = false;
            BOOL stairs_up = false;
            for(RouteStep *step in all_steps){
                if([[[step start] getBuildingName] isEqual:buildingName] && [[step path] pathType] == TYPE_STAIR){
                    if([[step start] getFloorNumber] > [[step end] getFloorNumber]){
                        stairs_down = true;
                    }else{
                        stairs_up = true;
                    }
                }
            }
            
            if(stairs_down){
                drawImageOnMap(context, _stairsDownImage, [[vec_c objectAtIndex:0] doubleValue]*25+[wp_cur x], [[vec_c objectAtIndex:1] doubleValue]*25+[wp_cur y], 35);
            }
            if(stairs_up){
                drawImageOnMap(context, _stairsUpImage, [[vec_c objectAtIndex:0] doubleValue]*25+[wp_cur x], [[vec_c objectAtIndex:1] doubleValue]*25+[wp_cur y], 35);
            }
            
        }
    }

}

- (Building*) determineBuildingFromPositionWithX:(float)x WithY:(float)y Threshold:(float)threshold{
    NSMutableArray *buildings = [_map buildings];
    
    Building *closest = nil;
    float closest_dist = 1e9;
    
    for(Building *building in buildings){
        if(![building selectable]) continue;
        
        float dist = [[[building getMainFloor] position]distanceTo:[[Waypoint alloc] initWithX:x withY:y]];
        if(dist < closest_dist){
            closest = building;
            closest_dist = dist;
        }
    }
    
    if(closest_dist > threshold)
        closest = nil;
    
    return closest;
}

@end

File: android/WATisRain/WATisRain/src/main/java/com/lucky/watisrain/map/MapView.java
package com.lucky.watisrain.map;

import android.content.Context;
import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.MenuItem;

import com.lucky.watisrain.Global;
import com.lucky.watisrain.R;
import com.lucky.watisrain.backend.MapFactory;
import com.lucky.watisrain.backend.RouteFinder;
import com.lucky.watisrain.backend.Util;
import com.lucky.watisrain.backend.data.Building;
import com.lucky.watisrain.backend.data.Map;
import com.lucky.watisrain.backend.data.Path;
import com.lucky.watisrain.backend.data.Route;
import com.lucky.watisrain.backend.data.RouteStep;
import com.lucky.watisrain.backend.data.Waypoint;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.List;

import uk.co.senab.photoview.PhotoView;
import uk.co.senab.photoview.PhotoViewAttacher;

/**
 * Class that draws the map. It handles reading and storing the Map object.
 */
public class MapView extends PhotoView {
	
	// Needed because PhotoView.getDisplayRect doesn't actually work
	public PhotoViewAttacher attacher;
	
	public HashMap<String, Bitmap> imgs;
	
	// Contains a TextView that displays directions
	public DirectionsView directionsView;
	
	public MenuItem clearBtn;
	
	public Map map;
	public RouteFinder routefinder;
	
	// Current route state
	String selectedBuilding1 = null;
	String selectedBuilding2 = null;
	Route route = null;

	public MapView(Context context, AttributeSet attrs) {
		super(context, attrs);

		AssetManager assetManager = context.getAssets();

		if(assetManager != null) {
			// Handle reading map file
			try {
				InputStream in = assetManager.open("locations.txt");
				map = MapFactory.readMapFromStream(in);
			} catch (IOException e) {
				Global.println(e);
			}
		} else {
			map = new Map();
		}
		
		routefinder = new RouteFinder(map);
		
		// Read bitmaps
		imgs = new HashMap<String, Bitmap>();
		imgs.put("default_location.png", BitmapFactory.decodeResource(getResources(), R.drawable.default_location));
		imgs.put("active_location.png", BitmapFactory.decodeResource(getResources(), R.drawable.active_location));
		imgs.put("stairs_up.png", BitmapFactory.decodeResource(getResources(), R.drawable.stairs_up));
		imgs.put("stairs_down.png", BitmapFactory.decodeResource(getResources(), R.drawable.stairs_down));
	}

	@Override
	protected void onDraw(Canvas canvas) {
		super.onDraw(canvas);
		Paint paint = new Paint();

		RectF displayRect = new RectF();

		if(attacher != null) {
			displayRect = attacher.getDisplayRect();
		}

		MapDraw mapdraw = new MapDraw(canvas, displayRect);
		
		// Draw all locations
		for(Building building : map.getBuildings()){
			
			if(!building.isSelectable()) continue;
			
			Waypoint pos = building.getMainFloor().getPostion();
			mapdraw.drawImageOnMap(imgs.get("default_location.png"),pos.getX(),pos.getY(),120);
		}
		
		// draw route
		if(route != null){
			
			List<RouteStep> all_steps = route.getRouteSteps();
			for(RouteStep step : all_steps){
				paint.setColor(Color.parseColor("#0070cf"));
				paint.setStrokeCap(Paint.Cap.ROUND);
				mapdraw.drawPathOnMap(step.getPath(), 8.0f, paint);
			}
		}
		
		// Draw active locations
		for(Building building : map.getBuildings()){
			Waypoint pos = building.getMainFloor().getPostion();
			
			if(building.getName().equals(selectedBuilding1) ||
			   building.getName().equals(selectedBuilding2)){
				
				mapdraw.drawImageOnMap(imgs.get("active_location.png"),pos.getX(),pos.getY(),120);
			}
		}
		
		
		
		// The following code handles drawing stairs icons.
		if(route != null){
			
			List<RouteStep> all_steps = route.getRouteSteps();
			
			// Get list of buildings we go through
			ArrayList<String> throughBuildings = new ArrayList<String>();
			throughBuildings.add(all_steps.get(0).getStart().getBuildingName());
			for(RouteStep step : all_steps) {
				String next_build = step.getEnd().getBuildingName();
				if(!throughBuildings.contains(next_build))
					throughBuildings.add(next_build);
			}
			
			// All waypoints that we go through
			ArrayList<Waypoint> throughWaypoints = new ArrayList<Waypoint>();
			for(RouteStep step : all_steps){
				throughWaypoints.addAll(step.getWaypoints());
			}
			// Filter duplicates
			throughWaypoints = new ArrayList<Waypoint>(new LinkedHashSet<Waypoint>(throughWaypoints));
			
			for(String buildingName : throughBuildings){
				int ix = throughWaypoints.indexOf(map.getBuildingByID(buildingName).getPosition());
				
				// Generate 3 waypoints to get our vectors from
				Waypoint wp_cur = throughWaypoints.get(ix);
				Waypoint wp_before;
				if(ix==0){
					wp_before = throughWaypoints.get(1);
				}else{
					wp_before = throughWaypoints.get(ix-1);
				}
				Waypoint wp_after;
				if(ix==throughWaypoints.size()-1){
					wp_after = throughWaypoints.get(throughWaypoints.size()-2);
				}else{
					wp_after = throughWaypoints.get(ix+1);
				}
				
				// Compute opposite vector
				double[] vec_c = Util.findOppositeVector(
						wp_before.getX()-wp_cur.getX(),
						wp_before.getY()-wp_cur.getY(),
						wp_after.getX()-wp_cur.getX(),
						wp_after.getY()-wp_cur.getY());
				
				
				// Determine whether we actually draw a stairs here
				boolean stairs_down = false;
				boolean stairs_up = false;
				for(RouteStep step : all_steps){
					if(step.getStart().getBuildingName().equals(buildingName)
							&& step.getPathType() == Path.TYPE_STAIR){
						if(step.getStart().getFloorNumber() > step.getEnd().getFloorNumber()){
							stairs_down = true;
						}else{
							stairs_up = true;
						}
					}
				}
				
				// Draw stair icons if needed
				if(stairs_down){
					mapdraw.drawImageOnMap(imgs.get("stairs_down.png"), (float)vec_c[0]*25+wp_cur.getX(), (float)vec_c[1]*25+wp_cur.getY(), 35);
				}
				if(stairs_up){
					mapdraw.drawImageOnMap(imgs.get("stairs_up.png"), (float)vec_c[0]*25+wp_cur.getX(), (float)vec_c[1]*25+wp_cur.getY(), 35);
				}
				
			}
		}
		
	}
	
	
	/**
	 * Called when user taps at location x,y on map
	 * 
	 * The units are relative to the image and between 0.0 and 1.0. That is, the center
	 * of the map is (0.5, 0.5)
	 */
	public void handleUserTap(float x, float y){
		
		if(clearBtn != null)
			clearBtn.setVisible(true);
		
		// Convert to map units
		// On map_full, it is (x * width, y * height). Then we perform scaling.
		float map_x = (x * Global.MAP_WIDTH / Global.MAP_ADJUST_SCALING) + Global.MAP_ADJUST_X;
		float map_y = (y * Global.MAP_HEIGHT / Global.MAP_ADJUST_SCALING) + Global.MAP_ADJUST_Y;
		
		Building closestBuilding = determineBuildingFromPosition(map_x, map_y, 70);
		
		// Reset route if user clicks one of existing endpoints, or something outside
		if(closestBuilding == null ||
				closestBuilding.getName().equals(selectedBuilding1) ||
				closestBuilding.getName().equals(selectedBuilding2)){
			clearRoute();
		}
		else if(selectedBuilding1 == null){
			selectedBuilding1 = closestBuilding.getName();
			directionsView.selectDestination(selectedBuilding1);
		}else{
			selectedBuilding2 = closestBuilding.getName();
			
			updateRoute();
		}
		
	}
	
	
	/**
	 * Clear any route or selected buildings, as well as text view
	 */
	public void clearRoute(){
		selectedBuilding1 = null;
		selectedBuilding2 = null;
		route = null;
		directionsView.unselectDestination();
		if(clearBtn != null)
			clearBtn.setVisible(false);
		invalidate();
	}
	
	
	// Assumes selectedBuilding[1,2] are the correct ones.
	private void updateRoute(){
		
		route = routefinder.findRoute(map.getBuildingByID(selectedBuilding1),
									  map.getBuildingByID(selectedBuilding2)).getContractedRoute();
		
		directionsView.generateDirectionsFromRoute(route);
	}
	
	
	/**
	 * Recalculate the route if applicable, possibly with a different global pathing
	 * value. If no route is selected, do nothing.
	 */
	public void recalculateRoute(){
		if(selectedBuilding1 == null || selectedBuilding2 == null) return;
		
		updateRoute();
		invalidate();
	}
	
	
	/**
	 * Given a set of map coordinates (x,y), return the building closest to
	 * it, or null if there is no building close by.
	 * 
	 * @param threshold maximum allowed distance in map-pixels
	 */
	private Building determineBuildingFromPosition(float x, float y, float threshold){
		
		List<Building> buildings = map.getBuildings();
		
		Building closest = null;
		float closest_dist = Float.MAX_VALUE;
		
		for(Building building : buildings){
			
			if(!building.isSelectable()) continue;
			
			float dist = (float)building.getMainFloor().getPostion().distanceTo(new Waypoint((int)x,(int)y));
			if(dist < closest_dist){
				closest = building;
				closest_dist = dist;
			}
		}
		
		if(closest_dist > threshold)
			closest = null;
		
		return closest;
	}

}
